thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/DetectionConfigTuner.java

```diff
@@ -0,0 +1,158 @@
+  public static final Set<String> TURNOFF_TUNING_COMPONENTS =
+      ImmutableSet.of("MIGRATED_ALGORITHM_FILTER", "MIGRATED_ALGORITHM", "MIGRATED_ALGORITHM_BASELINE");
+
+  private final DetectionConfigDTO detectionConfig;
+  private final DataProvider dataProvider;
+  private final DatasetConfigDTO datasetConfig;
+  private final String metricUrn;
+
+  public DetectionConfigTuner(DetectionConfigDTO config, DataProvider dataProvider) {
+    this.detectionConfig = config;
+    this.dataProvider = dataProvider;
+
+    Map<String, Object> yamlConfig = ConfigUtils.getMap(new org.yaml.snakeyaml.Yaml().load(config.getYaml()));
+
+    MetricConfigDTO metricConfig = dataProvider.fetchMetric(
+        MapUtils.getString(yamlConfig, PROP_METRIC),
+        MapUtils.getString(yamlConfig, PROP_DATASET));
+    Preconditions.checkNotNull(metricConfig, "metric not found");
+    this.datasetConfig = dataProvider.fetchDatasets(Collections.singletonList(metricConfig.getDataset()))
+        .get(metricConfig.getDataset());
+    Preconditions.checkNotNull(this.datasetConfig, "dataset not found");
+
+    this.metricUrn = MetricEntity.buildMetricUrn(MapUtils.getMap(yamlConfig, PROP_FILTERS), metricConfig.getId());
+  }
+
+  /**
+   * Returns the default(for new alert) or tuned configuration for the specified component
+   *
+   * @param componentKey the key used to reference the component within the detection
+   * @param componentProps previously tuned configuration along with user supplied yaml config
+   * @param startTime start time of the tuning window
+   * @param endTime end time of the tuning window
+   */
+  private Map<String, Object> getTunedSpecs(String componentKey, Map<String, Object> componentProps, long startTime, long endTime)
+      throws Exception {
+    // Instantiate tunable component
+    long configId = detectionConfig == null ? 0 : detectionConfig.getId();
+    String componentClassName = componentProps.get(PROP_CLASS_NAME).toString();
+    Map<String, Object> yamlParams = ConfigUtils.getMap(componentProps.get(PROP_YAML_PARAMS));
+    InputDataFetcher dataFetcher = new DefaultInputDataFetcher(dataProvider, configId);
+    Tunable tunable = instantiateTunable(componentClassName, yamlParams, dataFetcher);
+
+    // round to daily boundary
+    DateTimeZone timezone = DateTimeZone.forID(this.datasetConfig.getTimezone() == null ? DEFAULT_TIMEZONE : this.datasetConfig.getTimezone());
+    DateTime start = new DateTime(startTime, timezone).withTimeAtStartOfDay();
+    DateTime end =  new DateTime(endTime, timezone).withTimeAtStartOfDay();
+    Interval window = new Interval(start, end);
+
+    // TODO can you use componentProps instead? what is the difference btw component and componentSpec
+    Map<String, Object> existingComponentSpec =
+        detectionConfig.getComponents().containsKey(componentKey) ?
+            MapUtils.getMap(detectionConfig.getComponents(), componentKey) : Collections.emptyMap();
```

Comment:
should here be `MapUtils.getMap(detectionConfig.getComponentSpecs(), componentKey)  ` ? By calling `detectionConfig.getComponents()`, it will return a initilized  component object.